spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java

```diff
@@ -0,0 +1,312 @@
+import com.google.cloud.bigquery.TableDataWriteChannel;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.WriteChannelConfiguration;
+
+import org.springframework.cloud.gcp.bigquery.integration.BigQuerySpringMessageHeaders;
+import org.springframework.expression.EvaluationContext;
+import org.springframework.expression.Expression;
+import org.springframework.expression.common.LiteralExpression;
+import org.springframework.integration.expression.ExpressionUtils;
+import org.springframework.integration.expression.FunctionExpression;
+import org.springframework.integration.expression.ValueExpression;
+import org.springframework.integration.handler.AbstractReplyProducingMessageHandler;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHandlingException;
+import org.springframework.util.Assert;
+import org.springframework.util.StreamUtils;
+import org.springframework.util.concurrent.SettableListenableFuture;
+
+/**
+ * A {@link org.springframework.messaging.MessageHandler} which handles sending and
+ * loading files to a BigQuery table.
+ *
+ * @author Daniel Zou
+ * @since 1.2
+ */
+public class BigQueryFileMessageHandler extends AbstractReplyProducingMessageHandler {
+
+	private final BigQuery bigQuery;
+
+	private EvaluationContext evaluationContext;
+
+	private Expression datasetNameExpression;
+
+	private Expression tableNameExpression;
+
+	private Expression formatOptionsExpression;
+
+	private boolean autoDetectSchema = true;
+
+	private WriteDisposition writeDisposition = WriteDisposition.WRITE_APPEND;
+
+	private Duration timeout = Duration.ofMinutes(5);
+
+	private Duration jobPollInterval = Duration.ofSeconds(2);
+
+	private boolean sync = false;
+
+	public BigQueryFileMessageHandler(BigQuery bigQuery) {
+		Assert.notNull(bigQuery, "BigQuery client must not be null.");
+		this.bigQuery = bigQuery;
+
+		this.datasetNameExpression =
+				new FunctionExpression<Message>(
+						message -> message.getHeaders().get(BigQuerySpringMessageHeaders.DATASET_NAME));
+
+		this.tableNameExpression =
+				new FunctionExpression<Message>(
+						message -> message.getHeaders().get(BigQuerySpringMessageHeaders.TABLE_NAME));
+
+		this.formatOptionsExpression =
+				new FunctionExpression<Message>(
+						message -> message.getHeaders().get(BigQuerySpringMessageHeaders.FORMAT_OPTIONS));
+	}
+
+	@Override
+	protected void doInit() {
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
+	}
+
+	/**
+	 * Sets the SpEL {@link Expression} to evaluate to determine the dataset name.
+	 * @param datasetNameExpression the SpEL expression used to evaluate the dataset name
+	 */
+	public void setDatasetNameExpression(Expression datasetNameExpression) {
+		Assert.notNull(datasetNameExpression, "Dataset name expression can't be null.");
+		this.datasetNameExpression = datasetNameExpression;
+	}
+
+	/**
+	 * Sets the BigQuery dataset name to use.
```

Comment:
Mention in javadoc that it's mutually exclusive with `setDatasetNameExpression()`? Or that whichever is last will clobber. If you do it, then same for table name and format options.